File: orcastork/operators/context.py
"""The per-invocation ``OperatorContext`` and its change ``InvocationDelta``.

Every ``run(ctx)`` receives the current state (``store``), the currently-available
capabilities (``capabilities``), and a per-operator ``delta`` of what changed since *this*
operator last ran — so a rerun does incremental work instead of re-scanning.
"""

from __future__ import annotations

from collections.abc import Mapping
from contextlib import AbstractAsyncContextManager
from dataclasses import dataclass
from typing import Any

from ..aggregation.helpers import AggregationHelpers
from ..capabilities.base import Capability
from ..datapoints import BaseDataPoint, DataPointView
from ..exceptions import CapabilityUnavailableError
from ..ids import CapabilityId, Epoch, SessionId
from .effects import EffectGuard, EffectRecovery


@dataclass(frozen=True)
class InvocationDelta:
    added: frozenset[BaseDataPoint[Any]]  # new (type, value) identities
    updated: frozenset[BaseDataPoint[Any]]  # existing identities re-observed (last_retrieved bumped)
    newly_available_caps: frozenset[CapabilityId]  # capabilities that came online since the last run
    is_first_invocation: bool  # first run → `added` is the full current set


class CapabilityView:
    """Read view over the currently-available capabilities (resolve → preferred provider)."""

    def __init__(
        self,
        available: Mapping[CapabilityId, Capability] | None = None,
        preference: tuple[CapabilityId, ...] = (),
    ) -> None:
        self._available: dict[CapabilityId, Capability] = dict(available) if available is not None else {}
        self._preference = preference

    def available_ids(self) -> frozenset[CapabilityId]:
        return frozenset(self._available)

    def available_types(self) -> frozenset[type[Capability]]:
        return frozenset(type(capability) for capability in self._available.values())

    def is_available(self, capability_id: CapabilityId) -> bool:
        return capability_id in self._available

    def resolve[C: Capability](self, capability_type: type[C]) -> C | None:
        """The single preferred available provider of ``capability_type``, or None.

        Providers listed in the namespace's preference order win, in listed order; unlisted providers rank
        after every listed one and fall back to the stable alphabetical tie-break by id.

        Call the returned capability's action methods directly — they keep their real typed signatures,
        and each call audits itself (the framework wires the auditor when the capability activates).
        """
        matches = sorted(
            (cap for cap in self._available.values() if isinstance(cap, capability_type)),
            key=self._preference_rank,
        )
        return matches[0] if matches else None

    def _preference_rank(self, capability: Capability) -> tuple[int, CapabilityId]:
        if capability.capability_id in self._preference:
            return (self._preference.index(capability.capability_id), capability.capability_id)
        return (len(self._preference), capability.capability_id)

    def require[C: Capability](self, capability_type: type[C]) -> C:
        """Like :meth:`resolve`, but raise ``CapabilityUnavailableError`` when no provider is available.

        An operator that declared the capability in ``requires`` is only scheduled once it is
        available, so it can ``require`` the provider and use it without a ``None`` check.
        """
        capability = self.resolve(capability_type)
        if capability is None:
            raise CapabilityUnavailableError(f'no available provider for {capability_type.__name__}')
        return capability


@dataclass(frozen=True)
class OperatorContext:
    session_id: SessionId
    epoch: Epoch
    store: DataPointView
    capabilities: CapabilityView
    delta: InvocationDelta
    effects: EffectGuard  # claim/commit/revert gate for non-idempotent side effects (see operators/effects.py)
    aggregation: AggregationHelpers | None = None  # set only for aggregators (durable-write API)
    is_final: bool = False  # True only during the authoritative finalize pass; False for interim runs

    def latest[T: BaseDataPoint[Any]](self, data_point_type: type[T]) -> T | None:
        """The newest DataPoint of ``data_point_type`` by ``last_retrieved`` (ergonomic single-read)."""
        return self.store.latest(data_point_type)

    def once(
        self, effect_key: str, *, on_unknown: EffectRecovery = EffectRecovery.RERUN
    ) -> AbstractAsyncContextManager[bool]:
        """Guard a non-idempotent side effect — ``async with ctx.once('send-otp') as acquired:``.

        ``acquired`` is ``True`` iff this attempt owns running the effect; a clean exit commits
        the claim durably, a failing exit reverts it so a retry re-runs the effect.
        """
        return self.effects.once(effect_key, on_unknown=on_unknown)

File: typescript/src/orcastork/operators/context.ts
/**
 * The per-invocation `OperatorContext` and its change `InvocationDelta`.
 *
 * Every `run(ctx)` receives the current state (`store`), the currently-available capabilities
 * (`capabilities`), and a per-operator `delta` of what changed since *this* operator last ran — so a
 * rerun does incremental work instead of re-scanning.
 *
 * @module
 */

import type { AggregationHelpers } from '../aggregation/helpers.js';
import type { Capability, CapabilityClass } from '../capabilities/base.js';
import type { AnyDataPoint, DataPointClass, DataPointView } from '../datapoints/index.js';
import { CapabilityUnavailableError } from '../exceptions.js';
import type { CapabilityId, Epoch, SessionId } from '../ids.js';
import type { EffectBody, EffectGuard, OnceOptions } from './effects.js';

/**
 * What changed since *this* operator last ran — so a rerun does incremental work.
 *
 * A frozen plain object, not a validated model: the loop is its only constructor and builds one per
 * rerun-eligible operator on every pass, so validating DataPoints against types the engine itself
 * just computed would cost the hot path for nothing.
 */
export interface InvocationDelta {
  /** New `(type, value)` identities. */
  readonly added: ReadonlySet<AnyDataPoint>;

  /** Existing identities re-observed (`lastRetrieved` bumped). */
  readonly updated: ReadonlySet<AnyDataPoint>;

  /** Capabilities that came online since the last run. */
  readonly newlyAvailableCaps: ReadonlySet<CapabilityId>;

  /** First run → `added` is the full current set. */
  readonly isFirstInvocation: boolean;
}

/** What an {@link InvocationDelta} is built from; the sets are copied. */
export interface InvocationDeltaInit {
  readonly added: Iterable<AnyDataPoint>;
  readonly updated: Iterable<AnyDataPoint>;
  readonly newlyAvailableCaps: Iterable<CapabilityId>;
  readonly isFirstInvocation: boolean;
}

/** Build an {@link InvocationDelta} — the orchestrator's job, and no-one else's. */
export const InvocationDelta = (init: InvocationDeltaInit): InvocationDelta =>
  Object.freeze({
    added: new Set(init.added),
    updated: new Set(init.updated),
    newlyAvailableCaps: new Set(init.newlyAvailableCaps),
    isFirstInvocation: init.isFirstInvocation,
  });

/** Read view over the currently-available capabilities (resolve → preferred provider). */
export class CapabilityView {
  private readonly available: ReadonlyMap<CapabilityId, Capability>;
  private readonly preference: readonly CapabilityId[];

  public constructor(
    available: ReadonlyMap<CapabilityId, Capability> = new Map(),
    preference: readonly CapabilityId[] = [],
  ) {
    this.available = new Map(available);
    this.preference = [...preference];
  }

  public availableIds(): ReadonlySet<CapabilityId> {
    return new Set(this.available.keys());
  }

  public availableTypes(): ReadonlySet<CapabilityClass> {
    return new Set([...this.available.values()].map((capability) => capability.constructor as CapabilityClass));
  }

  public isAvailable(capabilityId: CapabilityId): boolean {
    return this.available.has(capabilityId);
  }

  /**
   * The single preferred available provider of `capabilityType`, or `null`.
   *
   * Providers listed in the namespace's preference order win, in listed order; unlisted providers
   * rank after every listed one and fall back to the stable alphabetical tie-break by id.
   *
   * Call the returned capability's action methods directly — they keep their real typed signatures,
   * and each call audits itself (the framework wires the auditor when the capability activates).
   */
  public resolve<C extends Capability>(capabilityType: CapabilityClass<C>): C | null {
    const matches = [...this.available.values()].filter(
      (capability): capability is C => capability instanceof capabilityType,
    );
    matches.sort((left, right) => this.compareByPreference(left, right));
    return matches[0] ?? null;
  }

  /**
   * Like {@link CapabilityView.resolve}, but throw `CapabilityUnavailableError` when no provider is
   * available.
   *
   * An operator that declared the capability in `requires` is only scheduled once it is available,
   * so it can `require` the provider and use it without a `null` check.
   */
  public require<C extends Capability>(capabilityType: CapabilityClass<C>): C {
    const capability = this.resolve(capabilityType);
    if (capability === null) {
      throw new CapabilityUnavailableError(`no available provider for ${capabilityType.name}`);
    }
    return capability;
  }

  private compareByPreference(left: Capability, right: Capability): number {
    const ranks = this.preferenceRank(left) - this.preferenceRank(right);
    if (ranks !== 0) {
      return ranks;
    }
    // Code-unit ordering, not locale ordering: the tie-break must not depend on the host.
    if (left.capabilityId === right.capabilityId) {
      return 0;
    }
    return left.capabilityId < right.capabilityId ? -1 : 1;
  }

  private preferenceRank(capability: Capability): number {
    const index = this.preference.indexOf(capability.capabilityId);
    return index === -1 ? this.preference.length : index;
  }
}

/** What an {@link OperatorContext} is built from — everything one invocation may see. */
export interface OperatorContextInit {
  readonly sessionId: SessionId;

  readonly epoch: Epoch;

  readonly store: DataPointView;

  readonly capabilities: CapabilityView;

  readonly delta: InvocationDelta;

  /** Claim/commit/revert gate for non-idempotent side effects (see `operators/effects.ts`). */
  readonly effects: EffectGuard;

  /** Aborted when the run is cut short — its timeout, or the session deadline. */
  readonly signal: AbortSignal;

  /** Set only for aggregators (the durable-write API); `null` for a gathering operator. */
  readonly aggregation?: AggregationHelpers | null;

  /** True only during the authoritative finalize pass; false for interim runs. */
  readonly isFinal?: boolean;
}

/**
 * What one invocation sees: the session's DataPoints, its capabilities, and what is new.
 *
 * Built by the orchestrator and frozen, never validated: the engine would only be checking itself,
 * once per launched run. `signal` has no Python counterpart — it is the port of the cancellation
 * `asyncio` gives Python for free: a promise cannot be cancelled, so a run that must stop (its
 * timeout fired, or the session deadline passed) is told through the signal, and anything genuinely
 * in flight inside `run` should be tied to it.
 */
export class OperatorContext {
  public readonly sessionId: SessionId;

  public readonly epoch: Epoch;

  /** Every DataPoint the session holds, not only this operator's dependencies. */
  public readonly store: DataPointView;

  public readonly capabilities: CapabilityView;

  /** What changed since this operator last ran. */
  public readonly delta: InvocationDelta;

  public readonly effects: EffectGuard;

  /** Aborted when this run is cut short; tie anything long-running to it. */
  public readonly signal: AbortSignal;

  /** The durable-write API — set only for aggregators. */
  public readonly aggregation: AggregationHelpers | null;

  /** True only during the authoritative finalize pass; false for interim runs. */
  public readonly isFinal: boolean;

  public constructor(init: OperatorContextInit) {
    this.sessionId = init.sessionId;
    this.epoch = init.epoch;
    this.store = init.store;
    this.capabilities = init.capabilities;
    this.delta = init.delta;
    this.effects = init.effects;
    this.signal = init.signal;
    this.aggregation = init.aggregation ?? null;
    this.isFinal = init.isFinal ?? false;
    Object.freeze(this);
  }

  /** The newest DataPoint of `dataPointType` by `lastRetrieved` (ergonomic single-read). */
  public latest<T extends AnyDataPoint>(dataPointType: DataPointClass<T>): T | null {
    return this.store.latest(dataPointType);
  }

  /**
   * Guard a non-idempotent side effect — `await ctx.once('send-otp', async (acquired) => { … })`.
   *
   * `acquired` is `true` iff this attempt owns running the effect; a clean exit commits the claim
   * durably, a body that throws reverts it so a retry re-runs the effect.
   */
  public async once(effectKey: string, body: EffectBody, options: OnceOptions = {}): Promise<boolean> {
    return await this.effects.once(effectKey, body, options);
  }
}
